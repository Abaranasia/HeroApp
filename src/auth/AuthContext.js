import { createContext } from "react";

export const AuthContext= createContext();
// lleva may√∫suclas porque es un componente, en este caso un high order component (apertura y cierre)