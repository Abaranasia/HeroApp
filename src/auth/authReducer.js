import { types } from "../types/types";

/** Reducer para tener bajo los estados de autenticación en la app
 * Toma los estados (login y logout) desde types
 * Este reducer estará compartido en toda la aplicación gracias a un context (AuthContext)
*/

export const authReducer = (state= {}, action) => {
    switch (action.type) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }
        case types.logout:
            return {
                logged: false
            }
        default:
            return state;
    }
}