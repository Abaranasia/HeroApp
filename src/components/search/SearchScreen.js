import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard'
import { getHerosByName } from '../../selectors/getHerosByName';
import { heroes } from '../../data/heroes';

export const SearchScreen = ( { history }) => {
/** Necesitamos un pequeño paquete para tratar los query strings y realizar las búsquedas de forma eficiente:
 * Query-string --> https://www.npmjs.com/package/query-string */
 

    const location= useLocation();
    //console.log(location.search) //nos permite obtener la información asociada a un parámetro de tipo search
    const {q= ''} = queryString.parse(location.search); // esta función nos permite parsear las búsquedas basadas en qs
    //console.log(q)

   
    
    const [ formValues, handleInputChange ] = useForm({  // Desestructuramos directamente la descripción en el formvalues
        searchText: q // tomamos la última búsqueda realizada
    })
    const {searchText} = formValues;


    const heroesFiltered= useMemo(() =>  getHerosByName(q), [q]); 
    // useMemo para evitar que se haga la búsqueda por cada letra introducida en el input
    //const heroesFiltered= getHerosByName(searchText); 
    //const heroesFilter= heroes; 
    
  
    const handleSearch = (e) => {
        e.preventDefault();
        //console.log(searchText)
        console.log("filtro: ",heroesFiltered);
        history.push(`?q=${searchText}`) //creamos el query string
        
    }
    
    return (
        <div>
            <h2>Search Screen</h2>
            <hr/>
            
            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <form onSubmit={ handleSearch }>
                        <input 
                            type="text"
                            placeholder="search hero"
                            className="form-control"
                            autoComplete="off"
                            name="searchText"
                            value={ searchText }
                            onChange={ handleInputChange }
                            />

                            <button
                                type="submit"
                                className="btn btn-primary"
                            >
                                Search
                            </button>
                    </form>
                </div>
                <div className="col-7 row">
                    <h4>Results</h4>
                    <hr/>
                    {
                        heroesFiltered.map ( hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        )) 
                    }

                </div>
                
            </div>
        </div>
    )
}
