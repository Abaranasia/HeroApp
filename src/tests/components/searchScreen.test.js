import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { SearchScreen } from '../../components/search/SearchScreen';

describe('Pruebas en searchScreen', () => {
    /** Probaremos: snapshot inicial, el handleSearch con history, 
     * llamada con un query-string que muestre resultados y otro sin resultados
     * Necesitaremos memoryRouter por el location y Route porque tenemos que trabajar con la ruta*/


    test('Debe mostrarse con valores por defecto', () => {
    //Primero probamos snapshop sin query-string y form vacío
        
        const wrapper= mount( 
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component= {SearchScreen} />
            </MemoryRouter>            
        )
    // Debemos especificar el mismo path en initialEntries o el snap se renderizará vacío por no estar en el sitio

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });


    test('Debe mostrarse con resultado e input con Batman', () => {
        //Primero probamos snapshop sin query-string y form vacío
            
            const wrapper= mount(           
                <MemoryRouter initialEntries={['/search?q=batman']}>
                    <Route path="/search" component= {SearchScreen} />
                </MemoryRouter>            
            )
        // Debemos especificar el mismo path en initialEntries o el snap se renderizará vacío por no estar en el sitio
    
            expect(wrapper).toMatchSnapshot();
            expect(wrapper.find('input').prop('value').toLowerCase()).toBe('batman');
            expect(wrapper.find('img').prop('alt').toLowerCase()).toBe('batman');
        });


    test('Debe mostrarse sin resultados pero input con Joker', () => {
        //Primero probamos snapshop sin query-string y form vacío
            
            const wrapper= mount(
                    <MemoryRouter initialEntries={['/search?q=joker']}>
                        <Route path="/search" component= {SearchScreen} />
                    </MemoryRouter>
            )
        // Debemos especificar el mismo path en initialEntries o el snap se renderizará vacío por no estar en el sitio
    
            expect(wrapper).toMatchSnapshot();
            expect(wrapper.find('input').prop('value').toLowerCase()).toBe('joker');
            expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no a hero with joker');
        });


    test('Debe llamar al push del history', () => {

        const historyMock = {
            push: jest.fn(),
        }
    
        const wrapper= mount(
                <MemoryRouter initialEntries={['/search?q=joker']}>
                    <Route 
                        path="/search" 
                        component= { () => <SearchScreen history= {historyMock} /> } 
                    />
                </MemoryRouter>
        ) // necesitamos mandarle el history

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        }); //accedemos al input y cambiamos su valor por batman

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })
        expect(wrapper).toMatchSnapshot();
        expect(historyMock.push).toHaveBeenCalledWith('?q=batman')
    });
});