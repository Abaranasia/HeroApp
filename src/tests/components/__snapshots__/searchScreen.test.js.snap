// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en searchScreen Debe llamar al push del history 1`] = `
<div>
  <h2>
    Search Screen
  </h2>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="search hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 row"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a hero with 
        joker
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en searchScreen Debe mostrarse con resultado e input con Batman 1`] = `
<div>
  <h2>
    Search Screen
  </h2>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="search hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 row"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-2"
        style={
          Object {
            "maxWidth": 250,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className=""
          >
            <a
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="./assets/heroes/dc-batman.jpg"
              />
            </a>
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                 
                Bruce Wayne
                 
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en searchScreen Debe mostrarse con valores por defecto 1`] = `
<div>
  <h2>
    Search Screen
  </h2>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="search hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 row"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en searchScreen Debe mostrarse sin resultados pero input con Joker 1`] = `
<div>
  <h2>
    Search Screen
  </h2>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="search hero"
          type="text"
          value="joker"
        />
        <button
          className="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 row"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a hero with 
        joker
      </div>
    </div>
  </div>
</div>
`;
